lib/Authentication.ts

```diff
@@ -49,6 +52,150 @@ export default class Authentication {
     this.factory = new CryptoFactory(options.cryptoSuites || [new RsaCryptoSuite()]);
   }
 
+  /**
+   * Signs the AuthenticationRequest with the private key of the Requester and returns the signed JWT.
+   * @param request well-formed AuthenticationRequest object
+   * @param responseDid DID of the requester.
+   */
+  public async signAuthenticationRequest (request: AuthenticationRequest): Promise<string> {
+    if (request.response_type !== 'id_token' || request.scope !== 'openid') {
+      throw new Error('Authentication Request not formed correctly');
+    }
+    const requesterDid = request.iss;
+    let key: PrivateKey | undefined;
+    key = this.getKey(requesterDid);
+    if (!key) {
+      throw new Error(`Could not find a key for ${requesterDid}`);
+    }
+
+    const token = new JwsToken(request, this.factory);
+    return token.sign(key);
+  }
+
+  /**
+   * Verifies signature on request and returns AuthenticationRequest.
+   * @param request Authentiation Request as a buffer or string.
+   */
+  public async verifyAuthenticationRequest (request: Buffer | string): Promise<AuthenticationRequest> {
+    // identity function in place of later signature verification
+    let jwsToken: JwsToken;
+    if (request instanceof Buffer) {
+      jwsToken = new JwsToken(request.toString(), this.factory);
+    } else {
+      jwsToken = new JwsToken(request, this.factory);
+    }
+    const keyId = jwsToken.getHeader().kid;
+    const keyDid = DidDocument.getDidFromKeyId(keyId);
+    const results = await this.resolver.resolve(keyDid);
+    const didPublicKey = results.didDocument.getPublicKey(keyId);
+    if (!didPublicKey) {
+      throw new Error('Could not find public key');
+    }
+    const publicKey = this.factory.constructPublicKey(didPublicKey);
+    const content = await jwsToken.verifySignature(publicKey);
+
+    const verifiedRequest: AuthenticationRequest = JSON.parse(content);
+    if (verifiedRequest.iss !== keyDid) {
+      throw new Error('Signing DID does not match issuer');
+    }

```

Comment:
Don't you also need to verify `scope` and `response_type` or are you accepting any kind of request and responding with a self-issued response regardless?

Reply:
I have in the AuthenticationRequest object that scope and response_type must be "openid" and "id_token" in response to Alan's previous comment, so we can't even call this function unless they are set to those string values.